### PR TITLE
[FIX] Restructuring

### PR DESCRIPTION
**Pull request by elephaint**
- [x] Put source files in src folder
- [x] Restructure methods adding Sparse methods to docs
- [x] Update CI to improve speed of tests
- [x] Update model performance tests to test more models
- [x] Docs only update on release or workflow dispatch, not on push to main
- [x] Move Numba functions to utils, perform eager compilation and micro optimization of lasso function.
- [x] Use `double` dtype everywhere; reconcilers are sensitive to low(er) precision, for now unify to `double` and maybe in the future turn everything to `single`

Lasso (ERM-reg / ERM-reg_bu) micro optimization before and after (the method is still slow as **** ):

Screenshot 2024-10-29 104009

Screenshot 2024-10-29 103932

### CODE REVIEW

File: .github/release-drafter.yml [hunk 1]
```diff
@@ -3,16 +3,14 @@ tag-template: 'v$NEXT_PATCH_VERSION'
 categories:
   - title: 'New Features'
     label: 'feature'
-  - title: 'Breaking Change'
-    label: 'breaking change'
```
Comment (jmoralez) [inline .github/release-drafter.yml:7]: Why were these removed?

File: .github/workflows/python-publish.yml [hunk 2]
```diff
@@ -28,8 +28,8 @@ jobs:
           python-version: '3.x'
       - name: Install dependencies
         run: |
-          python -m pip install --upgrade pip
-          pip install build
+          python -m pip install uv
+          uv pip install build
```
Comment (jmoralez) [inline .github/workflows/python-publish.yml:32]: I'm not sure this will work due to the venv. Since it's just a single package it's ok to leave it with pip. You could also use setup-uv and `uvx build` instead

File: action_files/test_models/src/data.py [hunk 3]
```diff
@@ -35,9 +35,9 @@ def get_data():
     spec = [
         ['Country'],
         ['Country', 'State'], 
-        ['Country', 'Purpose'], 
+        # ['Country', 'Purpose'],
```
Comment (jmoralez) [inline action_files/test_models/src/data.py:38]: Why were these removed?

File: action_files/test_models/src/evaluation.py [hunk 4]
```diff
@@ -32,7 +31,7 @@ def evaluate():
 
     eval_tags = {}
     eval_tags['Total'] = tags['Country']
-    eval_tags['Purpose'] = tags['Country/Purpose']
+    # eval_tags['Purpose'] = tags['Country/Purpose']
```
Comment (jmoralez) [inline action_files/test_models/src/evaluation.py:34]: Why was this removed?

File: action_files/test_models/src/models.py [hunk 5]
```diff
@@ -1,13 +1,15 @@
 import os
-import time
+# import time
```
Comment (jmoralez) [inline action_files/test_models/src/models.py:2]: If these are not used it's better to remove them

File: hierarchicalforecast/core.py [hunk 6]
```diff
@@ -244,17 +244,16 @@ def reconcile(self,
                 S_for_sparse = sparse.csr_matrix(S_df.sparse.to_coo())
             except AttributeError:
                 warnings.warn('Using dense S matrix for sparse reconciliation method.')
-                S_for_sparse = S_df.values.astype(np.float32)
+                S_for_sparse = S_df.values.astype(np.float64)
```
Comment (jmoralez) [inline hierarchicalforecast/core.py:247]: I'd add `copy=False` in these in case the array is already double

File: hierarchicalforecast/utils.py [hunk 7]
```diff
@@ -557,3 +565,202 @@ def samples_to_quantiles_df(samples: np.ndarray,
...
+                W[i, j] = W[j, i] = shrinkage * W[i, j]
+            else:
+                W[i, j] = W[j, i] = max(W[i, j], mint_shr_ridge)
+
+    return W
+
+# %% ../nbs/src/utils.ipynb 47
+# Lasso cyclic coordinate descent
+@njit("Array(float64, 1, 'C')(Array(float64, 2, 'C'), Array(float64, 1, 'C'), float64, int64, float64)", nogil=NUMBA_NOGIL, cache=NUMBA_CACHE, fastmath=NUMBA_FASTMATH, error_model="numpy")
+def _lasso(X: np.ndarray, y: np.ndarray,
```
Comment (jmoralez) [inline hierarchicalforecast/utils.py:728]: Is this function working as expected? I get either very big numbers of zeros.

File: .github/release-drafter.yml (hunk 1 above)
Comment (elephaint) [reply .github/release-drafter.yml:7]: Never mind, added them back.

File: action_files/test_models/src/data.py (hunk 3 above)
Comment (elephaint) [reply action_files/test_models/src/data.py:38]: Hierarchy needs to be strictly hierarchical

File: action_files/test_models/src/evaluation.py (hunk 4 above)
Comment (elephaint) [reply action_files/test_models/src/evaluation.py:34]: Hierarchy needs to be strictly hierarchical

File: hierarchicalforecast/core.py (hunk 6 above)
Comment (elephaint) [reply hierarchicalforecast/core.py:247]: Good point

File: hierarchicalforecast/utils.py (hunk 7 above)
Comment (elephaint) [reply hierarchicalforecast/utils.py:728]: This is a very valuable comment, thanks - this led to a rabbit hole uncovering that ERM-reg basically doesn't work at all, and it's also never tested (and was never tested before) - our unit test only tests ERM-reg_bu which does work. I looked back into old PRs and I don't think ERM-reg has ever worked.

To answer your questions:
- Zeros can be expected due to nan values present in the input data that ERM can't handle (I can think of how to handle these, but since ERM is essentially imho a fairly useless method as it's so computationally expensive I'm ok with this behaviour).
- If you'd perform a test without nan values you should see a P that's not full of zeros. 

I'll add a test for this and see if I can fix ERM-reg

Comment (elephaint) [reply hierarchicalforecast/utils.py:728]: Btw these things are the reason I added much more methods to the model_performance test, as our testing is very limited. But I think I'll add even more tests, seeing these kinds of bugs.

Comment (elephaint) [reply hierarchicalforecast/utils.py:728]: ERM-reg is basically garbage, I think we should leave it like this for now (old & new give same results)

**Comment by elephaint** [comment]
Test for lasso equivalence of methods (removed it from the code to keep it clean):

```
#| hide
# Lasso test old vs new
@njit
def _lasso_old(X: np.ndarray, y: np.ndarray, 
          lambda_reg: float, max_iters: int = 1_000,
          tol: float = 1e-4):
    # lasso cyclic coordinate descent
    n, feats = X.shape
    norms = (X ** 2).sum(axis=0)
    beta = np.zeros(feats, dtype=np.float32)
    beta_changes = np.zeros(feats, dtype=np.float32)
    residuals = y.copy()

    for it in range(max_iters):
        for i, betai in enumerate(beta):
            # is feature is close to zero, we 
            # continue to the next.
            # in this case is optimal betai= 0
            # print(beta)

            if abs(norms[i]) < 1e-8:
                continue
            xi = X[:, i]
            #we calculate the normalized derivative
            rho = betai + xi.flatten().dot(residuals) / norms[i] #(norms[i] + 1e-3)
            #soft threshold
            beta[i] = np.sign(rho) * max(np.abs(rho) - lambda_reg * n / norms[i], 0.)#(norms[i] + 1e-3), 0.)
            beta_changes[i] = np.abs(betai - beta[i])
            if beta[i] != betai:
                residuals += (betai - beta[i]) * xi
        if max(beta_changes) < tol:
            break
    #print(it)
    return beta


S = np.array([[1., 1., 1., 1.],
              [1., 1., 0., 0.],
              [0., 0., 1., 1.],
              [1., 0., 0., 0.],
              [0., 1., 0., 0.],
              [0., 0., 1., 0.],
              [0., 0., 0., 1.]])
h = 2
_y = np.array([10., 5., 4., 2., 1.])
y_bottom = np.vstack([i * _y for i in range(1, 5)])
y_hat_bottom_insample = np.roll(y_bottom, 1)
y_hat_bottom = np.vstack([i * np.ones(h) for i in range(1, 5)])
idx_bottom = [3, 4, 5, 6]

y_hat=S @ y_hat_bottom
y_insample=S @ y_bottom
idx_bottom=idx_bottom

n_hiers, n_bottom = S.shape

for method in ["reg", "reg_bu"]:
    for lambda_reg in [None, 0.1, 10, 1000]:
        for with_nans in [True, False]:
            y_hat_bottom_insample = np.roll(y_bottom, 1)
            if with_nans:
                y_hat_bottom_insample[:, 0] = np.nan
            y_hat_insample=S @ y_hat_bottom_insample
            y_insample=S @ y_bottom

            nan_idx = np.isnan(y_hat_insample).any(axis=0)
            y_insample = y_insample[:, ~nan_idx]
            y_hat_insample = y_hat_insample[:, ~nan_idx]
            h = min(y_hat.shape[1], y_hat_insample.shape[1])
            y_hat_insample = y_hat_insample[:, -h:] # shape (h, n_hiers)
            y_insample = y_insample[:, -h:]

            if method == 'reg':
                X = np.kron(S, y_hat_insample.T)
                z = y_hat_insample.reshape(-1)

                if lambda_reg is None:
                    lambda_reg = np.max(np.abs(X.T.dot(z)))
                else:
                    lambda_reg = lambda_reg            

                beta_old = _lasso_old(X, z, lambda_reg)
                beta_new = _lasso(X, z, lambda_reg, max_iters=1000, tol=1e-4) 
                np.testing.assert_allclose(beta_new, beta_old, atol=1e-5, rtol=1e-3)

            if method == 'reg_bu':
                X = np.kron(S, y_hat_insample.T)
                Pbu = np.zeros_like(S)
                Pbu[idx_bottom] = S[idx_bottom]
                z = y_hat_insample.reshape(-1) - X @ Pbu.reshape(-1)
                    
                if lambda_reg is None:
                    lambda_reg = np.max(np.abs(X.T.dot(z)))
                else:
                    lambda_reg = lambda_reg             

                beta_old = _lasso_old(X, z, lambda_reg)
                beta_new = _lasso(X, z, lambda_reg, max_iters=1000, tol=1e-4) 
                np.testing.assert_allclose(beta_new, beta_old, atol=1e-5, rtol=1e-3)
```

File: action_files/test_models/src/data.py (hunk 3 above)
Comment (jmoralez) [reply action_files/test_models/src/data.py:38]: I suggest we remove them instead of commenting them out.

File: action_files/test_models/src/evaluation.py (hunk 4 above)
Comment (jmoralez) [reply action_files/test_models/src/evaluation.py:34]: I suggest we remove them instead of commenting them out.

File: action_files/test_models/src/models.py [hunk 8]
```diff
@@ -17,8 +19,27 @@ def main():
...
                    MinTrace(method='mint_shrink'),
-                   MinTrace(method='ols')]
+                   MinTraceSparse(method='ols'),
+                   MinTraceSparse(method='wls_struct'),
+                   MinTraceSparse(method='wls_var'),
+                   OptimalCombination(method='ols'),
+                   OptimalCombination(method='wls_struct'),
+                   ERM(method='closed'),
+                #    ERM(method='reg'), # This is so insanely slow that we don't run it
+                #    ERM(method='reg_bu'), # This is so insanely slow that we don't run it
```
Comment (jmoralez) [inline action_files/test_models/src/models.py:41]: If these methods don't work we may as well remove them, people may be trying them out in a grid search or similar and just wasting time.

**Review by jmoralez — APPROVED** [review]
Non blocking comments.

File: action_files/test_models/src/evaluation.py (hunk 4 above)
Comment (elephaint) [reply action_files/test_models/src/evaluation.py:34]: Yeah, leaving these here as future-me wants to create additional tests when I'm introducing temporal reconciliation techniques, I probably want to try out strict- and non-strict hierarchical stuff then too.

For now my rationale was to get a reasonable test on every existing method

File: action_files/test_models/src/models.py (hunk 8 above)
Comment (elephaint) [reply action_files/test_models/src/models.py:41]: Agree, I'll be adding a deprecation notice in a future PR (or e.g. an exception when people try to run it with more than 10 timeseries)
